src/main/java/com/demo/classroom/Security/Service/JwtService.java

```diff
@@ -0,0 +1,134 @@
+    private final long REFRESH_EXPIRATION_TIME;
+
+    @Autowired
+    public JwtService(Environment env) {
+        this.SECRET_KEY = env.getProperty("JWT_SECRET_KEY");
+        this.JWT_EXPIRATION = 900000; 
+        this.REFRESH_EXPIRATION_TIME=3600000;
+    }
+    
+    public String extractUsername(String token) {
+        return extractClaim(token, Claims::getSubject);
+    }
+
+    public Long extractUserId(String token) {
+        Claims claims = extractAllClaims(token);
+        return claims.get("userId", Long.class);  
+    }
+
+    public List<String> extractRoles(String token) {
+        Claims claims = extractAllClaims(token);
+        return claims.get("roles", List.class);
+    }
+
+    public <T> T extractClaim(String token, Function<Claims, T> claimsResolver) {
+        final Claims claims = extractAllClaims(token);
+        return claimsResolver.apply(claims);
+    }
+    
+    private Claims extractAllClaims(String token){
+        return Jwts
+                .parserBuilder()
+                .setSigningKey(getSignInKey())
+                .build()
+                .parseClaimsJws(token)
+                .getBody();
+    }
+
+    public String generateToken(UserDetails userDetails){
+        Map<String, Object> claims = new HashMap<>();
+        claims.put("roles", userDetails.getAuthorities().stream()
+            .map(GrantedAuthority::getAuthority)
+            .collect(Collectors.toList()));
+        return generateToken(claims, userDetails);
+    }
+
+    public String generateToken(UserDetails userDetails, Long userID ){
+        Map<String, Object> claims = new HashMap<>();
+        claims.put("roles", userDetails.getAuthorities().stream()
+            .map(GrantedAuthority::getAuthority)
+            .collect(Collectors.toList()));
+        
+        claims.put("userId", userID);
+
+        return generateToken(claims, userDetails);
+    }
+
+    
+
+    public String refreshAccessToken(String refreshToken, UserDetails userDetails) {
+        if (isTokenValid(refreshToken, userDetails)) {
+            return generateToken(userDetails);
+        }
+        return null;
+    }
+
+    public String generateRefreshToken(UserDetails userDetails) {
+        Map<String, Object> claims = new HashMap<>();
+        return buildToken(claims, userDetails, REFRESH_EXPIRATION_TIME);
+    }
+
+    public String generateToken(Map<String, Object> extraClaims, UserDetails userDetails) {
+        return buildToken(extraClaims, userDetails, JWT_EXPIRATION);
+    }
+
+    private String buildToken(
+            Map<String, Object> extraClaims,
+            UserDetails userDetails,
+            long expiration
+    ) {
+        return Jwts
+                .builder()
+                .setClaims(extraClaims)
+                .setSubject(userDetails.getUsername())
+                .setIssuedAt(new Date(System.currentTimeMillis()))
+                .setExpiration(new Date(System.currentTimeMillis() + expiration))
```

Comment:
Instead of calling System.currentTimeMillis() multiple times, it's better to store the current timestamp in a variable and reuse it for both issuedAt and expiration to ensure consistency.

Reply:
Solved.